### PR TITLE
[FlashScript] initial upload of Python version.

### PR DESCRIPTION
**Pull request by andyblac**
please give through test, seems ok here.

### CODE REVIEW

File: tools/flash_shelly.py [hunk 1]
```diff
@@ -0,0 +1,499 @@
+import getopt, importlib, platform, urllib, json, logging, re
+from sys import argv, exit
+from os import path, remove
+from subprocess import Popen, PIPE
+from time import sleep
+from importlib import util
+
+"""
```
Comment (rojer) [inline tools/flash_shelly.py:8]: no need to make it a string. put it above imports and put `#!/usr/bin/env python3` as the first line

File: tools/flash_shelly.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
+import getopt, importlib, platform, urllib, json, logging, re
```
Comment (rojer) [inline tools/flash_shelly.py:1]: 1 import per line and sort them

File: tools/flash_shelly.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+from sys import argv, exit
```
Comment (rojer) [inline tools/flash_shelly.py:2]: don't use from foo import blah, import foo and use foo.blah

File: tools/flash_shelly.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+YELLOW='\033[1;93m' # Bright Yellow
+BLUE='\033[1;94m' # Bright Blue
+PURPLE='\033[1;95m' # Bright Purple
+NC='\033[0m' # Normal Color
+
+arch = platform.system()
+
+if not importlib.util.find_spec("zeroconf"):
+  print('Installing zeroconf...')
+  pipe = Popen('pip3 install zeroconf', shell=True, stdout=PIPE).stdout
```
Comment (rojer) [inline tools/flash_shelly.py:62]: `subprocess.check_output`

File: tools/flash_shelly.py [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+  print('Installing requests...')
+  pipe = Popen('pip3 install requests', shell=True, stdout=PIPE).stdout
+  out = (pipe.read())
+  import requests
+else:
+  import requests
+
+if not importlib.util.find_spec("packaging"):
+  print('Installing packaging...')
+  pipe = Popen('pip3 install packaging', shell=True, stdout=PIPE).stdout
```
Comment (rojer) [inline tools/flash_shelly.py:78]: extra dependency, let's not use it. write python version of the js function instead.

File: tools/flash_shelly.py [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+  out = (pipe.read())
+  from packaging import version
+else:
+  from packaging import version
+
+class MyListener:
+  global device_list, p_list
```
Comment (rojer) [inline tools/flash_shelly.py:85]: a global class variable? i didn't know you could do that. in any case, avoid global variables please, pass things around as needed or make them instance variables

File: tools/flash_shelly.py [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+               'switch2' : 'SHSW-21',
+               'switch25' : 'SHSW-25',
+               'shelly-plug-s' : 'SHPLG-S',
+               'dimmer1' : 'SHDM-1',
+               'rgbw2' : 'SHRGBW2',
+    }
+  return(options[type])
+
+def write_flash(device, lfw, dlurl, cfw_type, mode):
+  logger.info("\n" + WHITE + "write_flash" + NC)
```
Comment (rojer) [inline tools/flash_shelly.py:127]: you use both logger and prints for diagnostic output, use one or the other. my preference is logger.

File: tools/flash_shelly.py [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+  flashed = False
+  host = device.replace('.local','')
+  if cfw_type == 'homekit':
+    print("Downloading Firmware...")
+    logger.info('DURL: %s' % dlurl)
+    myfile = requests.get(dlurl)
+    open('shelly-flash.zip', 'wb').write(myfile.content)
```
Comment (rojer) [inline tools/flash_shelly.py:134]: this leaks an open file object.
use context managers on files, i.e.
```
with open(...) as f:
  ... do stuff ...
```

File: tools/flash_shelly.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+
+def probe_info(device, action, dry_run, silent_run, mode, exclude, exclude_device, forced_version, ffw):
+  flash = False
+  info = None           # firmware versions info
+  model = None          # device model
+  lfw = None            # latest firmware available
+  cfw = None            # current firmware on device
+  cfw_type = 'homekit'  # current firmware type
+  host = device
+  host = host.replace('.local','')
```
Comment (rojer) [inline tools/flash_shelly.py:202]: always put spaces after commas in argument lists

File: tools/flash_shelly.py [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+  if ((not opts and not args) and do_all == False) or (not args and do_all == False) or \
+     (args and do_all == True and exclude == False) or (not args and do_all == True and exclude == True):
+    usage()
+    exit(1)
+
+  try:
+    fp = urllib.request.urlopen("https://api.shelly.cloud/files/firmware")
+    global stock_release_info
+    stock_release_info = json.load(fp)
+    fp.close()
```
Comment (rojer) [inline tools/flash_shelly.py:467]: instead of closing explicitly, use context manager:

with urllib.request.urlopen(...) as fp:
 ...

File: tools/flash_shelly.py [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+  if mode == 'keep':
+    mode = cfw_type
+
+  if cfw_type == 'homekit':
+    type = info['app']
+    cfw = info['version']
+    if mode == 'homekit':
+      model = info['model'] if 'model' in info else shelly_model(type, mode)
+      for i in homekit_release_info:
+        if re.search(i[0][0], cfw):
```
Comment (rojer) [inline tools/flash_shelly.py:244]: shouldn't this be i[0]? i think i[0][0] is the first character of the regex

File: tools/flash_shelly.py [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+    fp = urllib.request.urlopen("https://api.shelly.cloud/files/firmware")
+    global stock_release_info
+    stock_release_info = json.load(fp)
+    fp.close()
+  except:
+    logger.warning("Failed to lookup version information")
+    exit(1)
+  try:
+    fp = urllib.request.urlopen("https://rojer.me/files/shelly/update.json")
+    global homekit_release_info
```
Comment (rojer) [inline tools/flash_shelly.py:473]: no global variables please, function arguments or instance variables.

File: tools/flash_shelly.py [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+  if do_all == True and exclude == True:
+    device_scan(args, action, do_all, dry_run, silent_run, mode, exclude, forced_version, ffw)
+
+
+
+
+if __name__ == '__main__':
+  app(argv)
+
+__all__ = ['app']
```
Comment (rojer) [inline tools/flash_shelly.py:499]: why is this needed?

**Review by rojer — CHANGES_REQUESTED** [review]

File: tools/flash_shelly.py (hunk 1 above)
Comment (andyblac) [reply tools/flash_shelly.py:8]: that stop windows from running script.

File: tools/flash_shelly.py (hunk 2 above)
Comment (andyblac) [reply tools/flash_shelly.py:1]: I normally code like above, but ok, np

File: tools/flash_shelly.py (hunk 3 above)
Comment (andyblac) [reply tools/flash_shelly.py:2]: ok, guess memory is not an issue.

File: tools/flash_shelly.py (hunk 5 above)
Comment (andyblac) [reply tools/flash_shelly.py:78]: no it not that, I use it here 
```
    elif (version.parse(cfw) < version.parse(lfw)) and ((cfw_type == 'homekit' and mode == 'homekit') or (cfw_type == 'stock' and mode == 'stock') or mode == "keep"):
```

File: tools/flash_shelly.py (hunk 6 above)
Comment (andyblac) [reply tools/flash_shelly.py:85]: I use it in device_scan , can't pass it there 'chicken and egg'.

Comment (andyblac) [reply tools/flash_shelly.py:85]: but I'll try rethink code.

File: tools/flash_shelly.py (hunk 7 above)
Comment (andyblac) [reply tools/flash_shelly.py:127]: you can't, that why logging states use print for normal running code.

File: tools/flash_shelly.py (hunk 11 above)
Comment (andyblac) [reply tools/flash_shelly.py:244]: nope.

File: tools/flash_shelly.py (hunk 13 above)
Comment (andyblac) [reply tools/flash_shelly.py:499]: runs the script from cmd line.

File: tools/flash_shelly.py (hunk 7 above)
Comment (andyblac) [reply tools/flash_shelly.py:127]: I use 2 levels of debug, we only have `info` and `debug` to use, so leaves nothing for standard logging.

File: tools/flash_shelly.py (hunk 6 above)
Comment (andyblac) [reply tools/flash_shelly.py:85]: this needs to be a global, as listener can not pass back data to previous def.

File: tools/flash_shelly.py (hunk 12 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/f65925f50d7ca9c6850d5735ace7621795b4e933

File: tools/flash_shelly.py (hunk 3 above)
Comment (andyblac) [reply tools/flash_shelly.py:2]:
```
from importlib import util
from sys import argv
```

these 2 are required, they can not be full imports. else modules is not found.

```
Andrews-MacBook-Pro:tools andrewblackburn$ python3 flash_shelly.py -la
Traceback (most recent call last):
  File "flash_shelly.py", line 68, in <module>
    if not importlib.util.find_spec("zeroconf"):
AttributeError: module 'importlib' has no attribute 'util'

Andrews-MacBook-Pro:tools andrewblackburn$ python3 flash_shelly.py -la
  File "flash_shelly.py", line 400
    def app(sys.argv):
               ^
SyntaxError: invalid syntax
```

Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/a39399332926dd2f642773d4765e5c36df3396fa
https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/5b81af09f69a18d8fde32f0891a5b4381ec425d2

File: tools/flash_shelly.py (hunk 2 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done.

File: tools/flash_shelly.py (hunk 4 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done 
https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/5c0a6611a21db479e0022bd105492863482a5220

File: tools/flash_shelly.py (hunk 8 above)
Comment (andyblac) [reply tools/flash_shelly.py:139]: like

```
    myfile = requests.get(dlurl)
    with open('shelly-flash.zip', 'wb') as f:
      write(myfile.content)
```

File: tools/flash_shelly.py (hunk 9 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/249ad1373b919b4a56bbb8355216dada8c2ce975

File: tools/flash_shelly.py (hunk 10 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/eeb8bae51752f3225a1425c7cec00adcb75675a3

File: tools/flash_shelly.py (hunk 11 above)
Comment (andyblac) [reply tools/flash_shelly.py:248]: oh you where correct close.
```
re.search(i[0][0]: <re.Match object; span=(0, 1), match='2'>
```
I'll tweak to
```
re.search(i[0]: <re.Match object; span=(0, 5), match='2.2.1'>
```

Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/2abe8439bc41b318b9192a0ee5c6aaacd23340c8

File: tools/flash_shelly.py (hunk 8 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/429770c217c5c98fe703eed838b2136debbd0185

File: tools/flash_shelly.py (hunk 13 above)
Comment (andyblac) [reply tools/flash_shelly.py:499]: it required to pass args and opts to getopt, they are stored in the import `argv `

File: tools/flash_shelly.py (hunk 5 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/5176422fad7ccf4ab1a55b93d76c7dea76b4cdaf

File: tools/flash_shelly.py (hunk 6 above)
Comment (andyblac) [reply tools/flash_shelly.py]: ok, reword to Instance Variables https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/ab391f6a2cd1fa5742591296f615544503758d27

File: tools/flash_shelly.py (hunk 7 above)
Comment (andyblac) [reply tools/flash_shelly.py:167]: https://docs.python.org/3/howto/logging.html

it states the following, which is what I am doing.
Screenshot 2020-10-07 at 01 03 50

Comment (andyblac) [reply tools/flash_shelly.py:167]: ok, I have managed to add a custom log level, tomorrow I'll replace print() with logging.

Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/d444400fdda05dc1ef91777762dfa621309da183

File: tools/flash_shelly.py (hunk 13 above)
Comment (andyblac) [reply tools/flash_shelly.py:542]: ie

```
if __name__ == "__main__":
   print("File one executed when ran directly")
else:
   print("File one executed when imported")
```

Comment (andyblac) [reply tools/flash_shelly.py:542]: I can remove `__all__ = ['app']` now we are not using `from xxx import yyyy`

Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/5286c45882aa179560b71b4ab5957a26d3397d00

File: tools/flash_shelly.py (hunk 1 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/215e8e65fb1b07da10257e558cfad61f55ace32e

Comment (andyblac) [reply tools/flash_shelly.py]: ok after reading up, it seems it won't effect windows at all, so I'll add it, thx

File: tools/flash_shelly.py (hunk 13 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/ae906910935bd41b57675857f7d42251938dad4c

File: tools/flash_shelly.py (hunk 1 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/ae906910935bd41b57675857f7d42251938dad4c

File: tools/flash_shelly.py (hunk 13 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/f9c90340cd36225fb2190d643fb8628d2b882543

Comment (andyblac) [reply tools/flash_shelly.py]: should be resolved now.

File: tools/flash_shelly.py (hunk 3 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/f9c90340cd36225fb2190d643fb8628d2b882543

File: tools/flash_shelly.py [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+logging.addLevelName(logging.TRACE, 'TRACE')
+logging.Logger.trace = functools.partialmethod(logging.Logger.log, logging.TRACE)
+logging.trace = functools.partial(logging.log, logging.TRACE)
+logging.basicConfig(format='%(message)s')
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+arch = platform.system()
+# Windows does not support acsii colours
+WHITE = '\033[1m' if not arch.lower().startswith('win') else ""
```
Comment (rojer) [inline tools/flash_shelly.py:62]: actually, this depends on terminal capabilities, not OS. maybe there's a way to figure it out.
but in any case, - avoid repetition where possible:

```
if not arch.lower().startswith('win'):
  ...
else:
  WHITE = RED = ... = NC = ""
```

File: tools/flash_shelly.py [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+    i = 0
+    for x in pp[1]:
+      c = pp[1][i]
+      if not c.isnumeric():
+        variant += c
+      else:
+        break
+      i += 1
+    varSeq = int(pp[1][i]) or 0
+  major = int(v[0])
```
Comment (rojer) [inline tools/flash_shelly.py:142]: more Pythonic way:

major, minor, patch = [int for e in v]

File: tools/flash_shelly.py [hunk 16]
```diff
@@ -0,0 +1,501 @@
...
+
+def write_flash(device, lfw, dlurl, cfw_type, mode):
+  logger.debug("\n" + WHITE + "write_flash" + NC)
+  flashed = False
+  host = device.replace('.local', '')
+  if cfw_type == 'homekit':
+    logger.info("Downloading Firmware...")
+    logger.debug('DURL: %s' % dlurl)
+    myfile = requests.get(dlurl)
+    with open('shelly-flash.zip', 'wb') as f:
```
Comment (rojer) [inline tools/flash_shelly.py:172]: you write this file only to read it back immediately.
if you need a file-like object backed by in-memory data, use io.BytesIO

File: tools/flash_shelly.py [hunk 17]
```diff
@@ -0,0 +1,501 @@
...
+  elif (vi1[3] != vi2[3]):
+    return True
+  elif (vi1[4] != vi2[4]):
+    return (vi1[4] > vi2[4])
+  else:
+    return False
+
+
+def write_flash(device, lfw, dlurl, cfw_type, mode):
+  logger.debug("\n" + WHITE + "write_flash" + NC)
```
Comment (rojer) [inline tools/flash_shelly.py:165]: using + to build strings is not pythonic.
since this is python 3, we can use the `f"..."` operator taht will perform variable interpolation, e.g.

logger.debug(f"\n{WHITE}write_flash{NC}")

much nicer and more readable

**Review by rojer — CHANGES_REQUESTED** [review]

File: tools/flash_shelly.py (hunk 14 above)
Comment (andyblac) [reply tools/flash_shelly.py:62]: doh ye np, don't know what I was thinking there.

File: tools/flash_shelly.py (hunk 15 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/da9181dd0e2545fbb3295e7b47a0cb81921fea92

File: tools/flash_shelly.py (hunk 16 above)
Comment (andyblac) [reply tools/flash_shelly.py:170]: it for update will io.Bytes be ok ?

so...
```
    myfile = requests.get(dlurl)
    with open('shelly-flash.zip', 'wb') as f:
      in_memory = io.BytesIO(f)
    logger.info("Now Flashing...")
    files = {'file': ('shelly-flash.zip', open(in_memory, 'rb'))}
    response = requests.post('http://%s/update' % device , files=files)
    logger.debug(response.text)
```
like that ?

Comment (andyblac) [reply tools/flash_shelly.py:170]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/58e3935f753894f457811da39e41ebb5146ce7d1

File: tools/flash_shelly.py (hunk 17 above)
Comment (andyblac) [reply tools/flash_shelly.py]: done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/ed7b9f018b7e77c98a3700db2e41a9753a03992d

File: tools/flash_shelly.py (hunk 15 above)
Comment (andyblac) [reply tools/flash_shelly.py]: hmmm did not work.
```
vi1: (<class 'int'>, <class 'int'>, <class 'int'>, '', 0)
vi2: (<class 'int'>, <class 'int'>, <class 'int'>, '', 0)
```

File: tools/flash_shelly.py (hunk 14 above)
Comment (andyblac) [reply tools/flash_shelly.py]: https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/758d4c4daeee3692a62120ec21df1d122c2d1736

**Comment by rojer** [comment]
great job!

**Comment by andyblac** [comment]
np, requires python 3.6, I think we should leave both versions up, bash scrpit can be used as fall back, I won't keep it updated with new features, but I'll try to keep basic functionality working.

**Comment by rojer** [comment]
no, i think we should leave just the python version. there isn't any real advantage, just maintenance burden.

**Comment by andyblac** [comment]
ok, np
